### ISSUE TITLE
Union type reflection failure confusion

### ISSUE THREAD

**Issue by bentayloruk** [fscheck/FsCheck#124]
*Thanks for FsCheck.  I'm going all out learning it on a new project :-)*

I've run into something I don't quite understand.  FsCheck is throwing the following exception for one of my types `ConditionExpr`.  I know this will happen when a type is not handled by a defaul FsCheck generator, but in this case, I'm not sure why this is failing.

The type in question is a DU that looks like this:

``` fsharp
type ConditionExpr = 
    | Fact        of Fact
    | Conditional of ConditionExpr * LogicalOperator * ConditionExpr
```

And the text exception is as follows (top snipped):

![screen shot 2015-07-27 at 09 36 24](https://cloud.githubusercontent.com/assets/189409/8901736/09abef84-3443-11e5-871b-51a7f58f3423.png)

I cloned FsCheck and debugged the failing test.  This is where I spot the thing I don't understand.  When reflecting on `ConditionExpr` `reflectObj` calls `isUnionType` on the `ConditionExpr` ([which it does here](https://github.com/fscheck/FsCheck/blob/master/src/FsCheck/ReflectArbitrary.fs#L574)), it returns `false`.  However, when I call the underlying FSharp reflection function `IsUnion` in FSI, it returns `true` (which is what I'd expect).

Sure this will be down to me somehow, but any pointers would be appreciated!

**Comment by mausch** [comment]
IIRC the reflection-based generator doesn't handle recursive types. See https://github.com/fscheck/FsCheck/blob/master/docs/Documentation.md#generating-recursive-data-types for instructions on how to define a generator for a recursive type.

**Comment by kurtschelfthout** [comment]
Hmm, no, it handles recursive types. The following works for me:

    type Fact = Fact
    type LogicalOperator = Op

    type ConditionExpr = 
        | Fact        of Fact
        | Conditional of ConditionExpr * LogicalOperator * ConditionExpr

    > Arb.generate<ConditionExpr> |> Gen.sample 10 1
    val it : ConditionExpr list =
      [Conditional (Conditional (Fact Fact,Op,Fact Fact),Op,Fact Fact)]

**Comment by kurtschelfthout** [comment]
I wonder if this is related to the changes to reflection for PCL. 

Maybe the line in Reflect.fs should read:

    let isUnionType ty = FSharpType.IsUnion(ty, true)

Are any of your types non-public?

**Comment by bentayloruk** [comment]
I think all the types are public.  However, I will check now.

**Comment by bentayloruk** [comment]
Right.  I had the solution open in VS2015 (recently installed).  Opened it in VS2013 and the tests run as expected without the exception.  Now I'm confused!

**Comment by kurtschelfthout** [comment]
Hm...F# 4.0 related problem?

**Comment by bentayloruk** [comment]
I'm referencing the FSharp.Core 3.1.2.5 Nuget package, but something in that area seems plausible.  4 was released to Nuget last week, so I might upgrade and see what happens.  Will do some investigating.

**Comment by kurtschelfthout** [comment]
Cool, thanks.

**Comment by bentayloruk** [comment]
Upgrading to 4.0 has given me the new problem of `FSharp.Core.sigdata not found alongside FSharp.Core` :-(

This could take a while!

**Comment by bentayloruk** [comment]
I don't know why this was not working in VS2015 with `3.1.2.5`, but now that I have the FSharp.Core `4.0.0.1`, it is working.  Therefore, I'm going to close, because I don't have more time right now to investigate.

*The sigdata/optdata issue I was having was due to getting an old, unlisted F# 4 package.  It looks like a possible NuGet issue, and I've raised it https://github.com/NuGet/Home/issues/1028.*
